homeassistant/components/dexcom/__init__.py

```diff
@@ -0,0 +1,87 @@
+"""The Dexcom integration."""
+import asyncio
+from datetime import timedelta
+import logging
+
+from pydexcom import AccountError, Dexcom, SessionError
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_UNIT_OF_MEASUREMENT, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import CONF_SERVER, DOMAIN, MG_DL, PLATFORMS, SERVER_OUS
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=180)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up configured Dexcom."""
+    hass.data[DOMAIN] = {}
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Dexcom from a config entry."""
+    try:
+        dexcom = await hass.async_add_executor_job(
+            Dexcom,
+            entry.data[CONF_USERNAME],
+            entry.data[CONF_PASSWORD],
+            entry.data[CONF_SERVER] == SERVER_OUS,
+        )
+    except AccountError:
+        return False
+    except SessionError:
+        raise ConfigEntryNotReady
+
+    entry.add_update_listener(update_listener)
+    if not entry.options:
+        entry.options = {CONF_UNIT_OF_MEASUREMENT: MG_DL}
+
+    async def async_update_data():
+        try:
+            return await hass.async_add_executor_job(dexcom.get_current_glucose_reading)
+        except SessionError as error:
+            raise UpdateFailed(error)
+
+    hass.data[DOMAIN][entry.entry_id] = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name=DOMAIN,
+        update_method=async_update_data,
+        update_interval=SCAN_INTERVAL,
+    )
+
+    await hass.data[DOMAIN][entry.entry_id].async_refresh()
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+    return unload_ok
+
+
+async def update_listener(hass, entry):
+    """Handle options update."""
+    hass.data[DOMAIN][entry.entry_id] = entry.options
```

Comment:
We're storing the update coordinator in hass.data under the entry_id key. We don't need to store the options. Those are available on the config entry.

We probably want to reload the config entry when the unit of measurement changes so that the entities use the new unit.